File: domains/knowledge-update/update_knowledge.py
#!/usr/bin/env python3
"""Control plane for data-analysis knowledge module updates."""

from __future__ import annotations

import argparse
import json
import subprocess
import sys
from pathlib import Path
from typing import Any


ROOT = Path(__file__).resolve().parents[2]
REGISTRY = ROOT / "domains" / "knowledge-update" / "registry.json"


def discover_modules() -> list[dict[str, Any]]:
    modules = []
    for module_path in sorted((ROOT / "domains").glob("*/module.json")):
        modules.append(json.loads(module_path.read_text(encoding="utf-8")))
    return modules


def write_registry(modules: list[dict[str, Any]]) -> None:
    payload = {
        "modules": [
            {
                "id": module["id"],
                "path": str((ROOT / "domains" / module["id"] / "module.json").relative_to(ROOT)),
                "serves": module.get("serves", []),
            }
            for module in modules
        ]
    }
    REGISTRY.write_text(json.dumps(payload, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")


def run_status(modules: list[dict[str, Any]]) -> int:
    exit_code = 0
    for module in modules:
        check = module.get("check", {})
        if check.get("type") != "script" or not check.get("cmd"):
            print(f"{module['id']}: unknown (no script check)")
            exit_code = 1
            continue
        result = subprocess.run(check["cmd"], cwd=ROOT, shell=True, text=True, capture_output=True)
        line = result.stdout.strip() or result.stderr.strip()
        try:
            payload = json.loads(line)
            status = payload.get("status", "unknown")
            messages = "; ".join(payload.get("messages", []))
        except json.JSONDecodeError:
            status = "unknown"
            messages = line
        print(f"{module['id']}: {status}" + (f" - {messages}" if messages else ""))
        if status != "fresh":
            exit_code = 1
    return exit_code


def run_update(modules: list[dict[str, Any]], target: str) -> int:
    selected = modules if target == "all" else [module for module in modules if module["id"] == target]
    if not selected:
        print(f"unknown module: {target}", file=sys.stderr)
        return 2

    exit_code = 0
    for module in selected:
        update = module.get("update", {})
        if update.get("type") == "script":
            result = subprocess.run(update["cmd"], cwd=ROOT, shell=True)
            if result.returncode != 0:
                exit_code = result.returncode
        elif update.get("type") == "agent":
            print(f"{module['id']}: agent update required")
            print(f"procedure: {update.get('procedure')}")
            links = module.get("doc_links", [])
            if links:
                print("doc_links:")
                for link in links:
                    print(f"  - {link}")
        else:
            print(f"{module['id']}: unknown update type", file=sys.stderr)
            exit_code = 1
    return exit_code


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    sub = parser.add_subparsers(dest="command", required=True)
    sub.add_parser("status")
    update = sub.add_parser("update")
    update.add_argument("target")
    sub.add_parser("register")
    args = parser.parse_args(argv)

    modules = discover_modules()
    if args.command == "register":
        write_registry(modules)
        print(f"registered {len(modules)} modules")
        return 0
    if args.command == "status":
        return run_status(modules)
    if args.command == "update":
        return run_update(modules, args.target)
    return 2


if __name__ == "__main__":
    raise SystemExit(main())

File: domains/knowledge-update/update_knowledge.ts
#!/usr/bin/env node
import { execSync, spawnSync } from "node:child_process";
import { readdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const root = join(dirname(fileURLToPath(import.meta.url)), "..", "..");
const registry = join(root, "domains", "knowledge-update", "registry.json");

interface ModuleSpec {
  id: string;
  serves?: string[];
  update?: { type?: string; cmd?: string; procedure?: string };
  check?: { type?: string; cmd?: string };
  doc_links?: string[];
}

function discoverModules(): ModuleSpec[] {
  return readdirSync(join(root, "domains"), { withFileTypes: true })
    .filter((entry) => entry.isDirectory())
    .map((entry) => join(root, "domains", entry.name, "module.json"))
    .filter((path) => {
      try {
        readFileSync(path);
        return true;
      } catch {
        return false;
      }
    })
    .sort()
    .map((path) => JSON.parse(readFileSync(path, "utf8")) as ModuleSpec);
}

function writeRegistry(modules: ModuleSpec[]): void {
  const payload = {
    modules: modules.map((module) => ({
      id: module.id,
      path: `domains/${module.id}/module.json`,
      serves: module.serves ?? []
    }))
  };
  writeFileSync(registry, JSON.stringify(payload, null, 2) + "\n", "utf8");
}

function runStatus(modules: ModuleSpec[]): number {
  let exitCode = 0;
  for (const module of modules) {
    const check = module.check ?? {};
    if (check.type !== "script" || !check.cmd) {
      console.log(`${module.id}: unknown (no script check)`);
      exitCode = 1;
      continue;
    }
    const result = spawnSync(check.cmd, { cwd: root, shell: true, encoding: "utf8" });
    const line = (result.stdout || result.stderr).trim();
    let status = "unknown";
    let messages = line;
    try {
      const payload = JSON.parse(line);
      status = payload.status ?? "unknown";
      messages = (payload.messages ?? []).join("; ");
    } catch {
      status = "unknown";
    }
    console.log(`${module.id}: ${status}${messages ? ` - ${messages}` : ""}`);
    if (status !== "fresh") exitCode = 1;
  }
  return exitCode;
}

function runUpdate(modules: ModuleSpec[], target: string): number {
  const selected = target === "all" ? modules : modules.filter((module) => module.id === target);
  if (selected.length === 0) {
    console.error(`unknown module: ${target}`);
    return 2;
  }
  let exitCode = 0;
  for (const module of selected) {
    const update = module.update ?? {};
    if (update.type === "script" && update.cmd) {
      try {
        execSync(update.cmd, { cwd: root, stdio: "inherit", shell: "/bin/sh" });
      } catch (error: any) {
        exitCode = error.status ?? 1;
      }
    } else if (update.type === "agent") {
      console.log(`${module.id}: agent update required`);
      console.log(`procedure: ${update.procedure}`);
      if ((module.doc_links ?? []).length > 0) {
        console.log("doc_links:");
        for (const link of module.doc_links ?? []) console.log(`  - ${link}`);
      }
    } else {
      console.error(`${module.id}: unknown update type`);
      exitCode = 1;
    }
  }
  return exitCode;
}

function main(argv = process.argv.slice(2)): number {
  const [command, target] = argv;
  const modules = discoverModules();
  if (command === "register") {
    writeRegistry(modules);
    console.log(`registered ${modules.length} modules`);
    return 0;
  }
  if (command === "status") return runStatus(modules);
  if (command === "update" && target) return runUpdate(modules, target);
  console.error("usage: update_knowledge.ts <status|register|update id|update all>");
  return 2;
}

if (import.meta.url === `file://${process.argv[1]}`) {
  process.exitCode = main();
}
